File: home_guard/tests/fake_remote_store.py
"""An in-memory fake conforming to crdt_sync's RemoteStore protocol, for tests."""

from __future__ import annotations

from crdt_sync import RemoteSyncError


class FakeRemoteStore:
    """Minimal in-memory stand-in; raises RemoteSyncError when told to fail."""

    def __init__(self, *, fail: bool = False) -> None:
        self._files: dict[str, str] = {}
        self.fail = fail

    def list_directory(self, path: str) -> list[str]:
        if self.fail:
            msg = "fake outage"
            raise RemoteSyncError(msg)
        prefix = f"{path}/"
        return [k[len(prefix) :] for k in self._files if k.startswith(prefix)]

    def get_file_text(self, path: str) -> str | None:
        if self.fail:
            msg = "fake outage"
            raise RemoteSyncError(msg)
        return self._files.get(path)

    def put_file_text(self, path: str, text: str, *, message: str = "") -> None:
        if self.fail:
            msg = "fake outage"
            raise RemoteSyncError(msg)
        self._files[path] = text

    def delete_file(self, path: str, *, message: str = "") -> None:
        if self.fail:
            msg = "fake outage"
            raise RemoteSyncError(msg)
        self._files.pop(path, None)

    def can_access_remote(self) -> bool:
        return not self.fail

    def close(self) -> None:
        return None

File: app/test/fake_remote_store.dart
import 'package:crdt_sync/crdt_sync.dart';

/// An in-memory `RemoteStore`: a flat path -> text map.
class MemRemote implements RemoteStore {
  /// Files by logical path.
  final Map<String, String> files = {};

  /// Every path written, in order, so a test can assert "wrote nothing".
  final List<String> puts = [];

  @override
  Future<List<String>> listDirectory(String path) async {
    final prefix = '$path/';
    return files.keys
        .where((k) => k.startsWith(prefix))
        .map((k) => k.substring(prefix.length).split('/').first)
        .toSet()
        .toList();
  }

  @override
  Future<String?> getFileText(String path) async => files[path];

  @override
  Future<void> putFileText(
    String path,
    String text, {
    required String message,
  }) async {
    files[path] = text;
    puts.add(path);
  }

  @override
  Future<void> deleteFile(String path, {String? message}) async =>
      files.remove(path);

  @override
  Future<bool> canAccessRemote() async => true;

  @override
  void close() {}
}

/// A store whose every call fails, as an unreachable network would.
class BrokenRemote implements RemoteStore {
  @override
  Future<List<String>> listDirectory(String path) async =>
      throw RemoteSyncError('offline');

  @override
  Future<String?> getFileText(String path) async =>
      throw RemoteSyncError('offline');

  @override
  Future<void> putFileText(
    String path,
    String text, {
    required String message,
  }) async => throw RemoteSyncError('offline');

  @override
  Future<void> deleteFile(String path, {String? message}) async =>
      throw RemoteSyncError('offline');

  @override
  Future<bool> canAccessRemote() async => false;

  @override
  void close() {}
}
